---
title: "Untitled"
author: "Nechama Wieder"
date: "2024-05-28"
output: html_document
---
Nechama Tatz-Wieder DPhil Clinical Medicine Thesis 2024

Chapter 4 - Investigating how 5’UTRs vary by gene tolerance to LoF
low LOEUF = low tolerance to inactivation ie more likely to be disease causing, also haploinsufficient
high LOEUF = high tolerance to inactivation ie less likely to be disease causing

#Data gnomAD V2
downloaded from https://gnomad.broadinstitute.org/downloads >> Under heading of “Constraint” >> “pLoF Metrics by Transcript”, file = "gnomad.v2.1.1.lof_metrics.by_transcript.txt.bgz"


#load libraries
```{r}
library(ggplot2)
library(splitstackshape)
library(stringr)
library(dplyr)
library(tidyverse)
```


#Prepare gnomAD Data 
```{r message=FALSE, error=FALSE, warning=FALSE}
gnomad1 <- read.table("gnomad.v2.1.1.lof_metrics.by_transcript.txt.bgz", fill = TRUE, header = TRUE)

#from nicky
#LOEUF has been calculated for all transcripts in GRCh37. MANE is on GRCh38 and hence many MANE transcripts are not represented in the LOEUF file. To get around this, we can take a single LOEUF score per gene - easiest way to do this is to take the value assigned to the canonical transcript (marked in the file) for every gene. Canonical transcripts are not perfect, this is why we use MANE wherever we can, but for this purpose they are a good proxy per gene.

#restrict to canonical, rebin (make new deciles), keep highest decile where genes are duplicated (ie >1 transcript for a gene as i want 1 trans: 1 gene)
gnomad <- gnomad1 %>% filter(canonical == "true")
gnomad <- gnomad[!is.na(gnomad$oe_lof_upper), ] #remove those with oe_lof_upper = na
#how many unique gene names are there?
#un1 <- unique(gnomad$gene) #19155, 42  genes entered >1
#for the genes >1, take the lower oe_lof_upper
gnomad_2 <- gnomad %>% group_by(gene) %>% slice(which.min(oe_lof_upper))#19155

### rebin unique canonical transcripts
gnomad_2$decile <- ntile(gnomad_2$oe_lof_upper, 10)

#subset gn_final to just the ones i want:
gn_final1 <- subset(gnomad_2, select=c("gene", "transcript", "pLI", "decile", "oe_lof_upper", "oe_lof_upper_bin", "oe_lof_upper_rank", "no_lofs", "obs_lof","transcript_type"))
```

#1. 5'UTR Lengths 
```{r message=FALSE, error=FALSE, warning=FALSE}
#bring in MANE 5' UTR length
length_exon <- read.table("length_introns.txt")

#####merge by gene id called transcript type in gnomad data
length_exon <- cSplit(length_exon, "gene_id", ".") 
names(gn_final1)[10]<-"gene_id"
names(length_exon)[6]<-"gene_id"
gnomad_len <- merge(length_exon, gn_final1, by = "gene_id") 
#17337

####stats - length differences in bottom and top 2 deciles

bot <- gnomad_len %>% filter(decile<3)
mean(bot$UTR_length)#268.8319

top <- gnomad_len %>% filter(decile>8)
mean(top$UTR_length)#162.3923
bot2 <- bot$UTR_length
top2 <- top$UTR_length
wilcox.test(bot2, top2)$p.value
#p=8.965308e-145
#significant

####plot
gnomad_len$decile <- gnomad_len$decile*10

ggplot(gnomad_len, aes(x=factor(decile), y=UTR_length)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("UTR Length (bp)") + geom_hline(yintercept = 202,linetype = "dotted", size=0.8) + theme(text = element_text(size = 23)) + scale_y_continuous(limits = c(0, 1500))


library(tidyverse)
library(ggthemes)  # for a mapping theme
library(ggalt)  # for custom map projections
library(ggrepel)  # for annotations
library(viridis)  # for nice colours

#remotes::install_github('jorvlan/raincloudplots')
#library(raincloudplots)

#source("https://gist.githubusercontent.com/benmarwick/2a1bb0133ff568cbe28d/raw/fb53bd97121f7f9ce947837ef1a4c65a73bffb3f/geom_flat_violin.R") - geom_flat_violin
#here dots are grey and boxplot purple. 
ggplot(data = gnomad_len, aes(x=factor(decile), y = UTR_length, fill = decile)) +   geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8, fill="#793A92") +     geom_point(aes(y = UTR_length, fill = decile), position = position_jitter(width = 0.15), size = 1, alpha = 0.1, color="grey48") +     geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.5, fill="#793A92") +     labs(x = "LEOUF Decile", y = "UTR Length (bp)") + coord_cartesian(ylim=c(0, 2000)) + guides(fill = FALSE, color = FALSE) +theme_light()  + theme(text = element_text(size = 20)) 

#export data as table
tab_len <- gnomad_len %>% group_by(decile) %>% summarize(mean_length=mean(UTR_length), median_length=median(UTR_length),range_length=paste(min(UTR_length),max(UTR_length), sep="-"), sd_length=sd(UTR_length))
#write_xlsx(tab_len, "2A_gnlen.xlsx") 
```

#2. 5'UTR Introns 
```{r message=FALSE, error=FALSE, warning=FALSE}
#1. plot diff between 0 and 1+ introns across LEOUF
gn_int<- gnomad_len %>% mutate(intron_cutoff = intron_final>0)
gn_int2 <- gn_int %>% group_by(decile) %>% dplyr::count(intron_cutoff)

total_tab <- as.data.frame(table(gnomad_len$decile))
total <- total_tab$Freq
#get percentages
tot <- rep(total, each=2)
gn_int2$total <- tot
gn_int2$perc <- (gn_int2$n/gn_int2$total) *100

#prepare for plot
#change intron true/false into meaningful
gn_int2 <- gn_int2 %>% mutate(intron_cutoff=str_replace(intron_cutoff, "TRUE", "1+ Intron"))
gn_int2 <- gn_int2 %>% mutate(intron_cutoff=str_replace(intron_cutoff, "FALSE", "0 Intron"))

ggplot(gn_int2, aes(x=factor(decile), y = perc, fill = intron_cutoff)) + geom_bar(position="dodge", stat="identity") + theme_light()   + xlab("LOEUF Decile") + ylab("Percentage") + scale_fill_manual(values = c("0 Intron" = "#BFACC8", "1+ Intron" = "#2B4595")) + theme(legend.title=element_blank()) + theme(text = element_text(size = 20))  #+ theme(legend.position="none") 



#2. Are the differences between 0 intron and >1 intron across deciles statistically significant?
#####stats
#make matrix
zero <- gn_int2 %>% filter(intron_cutoff=="0 Intron")
int <- gn_int2 %>% filter(intron_cutoff=="1+ Intron")

bot_z <- zero %>% filter(decile<30)
botz_genes <- sum(bot_z$n)
botz_tot <- sum(bot_z$total)

top_z <- zero %>% filter(decile>80)
top_zgenes <- sum(top_z$n)
topz_tot <- sum(top_z$total)

#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene <- c(botz_tot, topz_tot)
#how many genes in those deciles with 1+ intron
one_int <- c(botz_genes, top_zgenes)

int_matrix <- data.frame(zero="",int=one_int, total=tot_dec_gene)
int_matrix$zero <- int_matrix$total - int_matrix$int
#get percentages so can report in paper
int_matrix$perc_z <- (int_matrix$zero / int_matrix$total) *100
int_matrix$perc_o <- (int_matrix$int / int_matrix$total) *100
#do chi-square
test_int <- subset(int_matrix, select = c(zero, int))
chisq.test(test_int)$p.value

#0.185136

```
#3. uAUG/uORF/oORF/Start-stops
```{r message=FALSE, error=FALSE, warning=FALSE}
#1. get orf per gene data 
orf_gene <- read.table("orf_gene.txt")

#get gene_id
orf2 <- merge(orf_gene, length_exon, by="transcript")


#2. merge with gnomad
orf_gn <- merge(gn_final1, orf2, by="gene_id")

#3. aggregate to get totals of each uaug type per decile
orf_gn2 <- orf_gn %>% group_by(decile,orf_type) %>% summarise(n()) 
#n here is number of genes with at least 1 of that orf type


#3. get total per decile and append
#total <- data.frame(table(orf_gn$decile))
#this total is gnomad genes with uorfs but i want all gnomad genes

#use gnomad length for all totals - these are all gnomad genes which have a 5utr in mane
total <- data.frame(table(gnomad_len$decile))
totals <- total$Freq
totals2 <- rep(totals, each=3)
orf_gn2$total <- totals2

#percentage
orf_gn2$perc <- (orf_gn2$`n()`/orf_gn2$total) *100

#change all start-stop to Start-Stop
orf_gn2$orf_type <- as.character(orf_gn2$orf_type)
orf_gn2[orf_gn2 == "start-stop"] <- "Start-Stop"

###plot

orf_gn2$orf_type <- factor(orf_gn2$orf_type, levels = c("uORF", "oORF", "Start-Stop"))
orf_gn2$decile <- orf_gn2$decile*10 

#add mean lines 
table(orf_gene$orf_type)
#uo - 6461/18764 == 34.4%
#oorf - 2806/18764 == 15%
#ss - 940/18764 == 5%

ggplot(orf_gn2, aes(x=factor(decile), y=perc, fill=orf_type)) + geom_bar(position="dodge", stat="identity") + theme_light() + xlab("LOEUF Decile") + ylab("Percentage") + scale_fill_manual(values = c("uORF" = "NAVY", "oORF" = "palevioletred1", "Start-Stop" = "deepskyblue1"))  + geom_hline(yintercept = 34.4, colour="NAVY",linetype = "dotted") + geom_hline(yintercept = 15, colour="palevioletred1",linetype = "dotted") + geom_hline(yintercept =5, colour="deepskyblue1",linetype = "dotted")  + theme(legend.title=element_blank()) + theme(text = element_text(size = 20)) #+ theme(legend.position = "none")  

#########STATS
####uORFs
uo <- orf_gn2 %>% filter(orf_type=="uORF")
###stats for bottom and top 2 deciles start-stops between ahving 0 or 1 uorfs
bot_uo <-uo %>% filter(decile<30) 
top_uo <-  uo %>% filter(decile>80) 

#total of genes in bot and top 2 deciles
tot_bot_u <- sum(bot_uo$total)#3695
tot_top_u <- sum(top_uo$total)#2789

#total genes with uo in bot and top
botuo <- sum(bot_uo$`n()`)
topuo<- sum(top_uo$`n()`)

#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene_u <- c(tot_bot_u, tot_top_u)
#how many genes in those deciles with 1+ uo
one_uo <- c(botuo, topuo)

uo_matrix <- data.frame(zero="",uo=one_uo, total=tot_dec_gene_u)
uo_matrix$zero <- uo_matrix$total - uo_matrix$uo
#get percentages so can report in paper
uo_matrix$perc_z <- (uo_matrix$zero / uo_matrix$total) *100
uo_matrix$perc_o <- (uo_matrix$uo / uo_matrix$total) *100
#do chi-square
test_uo <- subset(uo_matrix, select = c(zero, uo))
chisq.test(test_uo)$p.value
#p=2.087572e-51
#sig

####start-stops
ss <- orf_gn2 %>% filter(orf_type=="Start-Stop")
###stats for bottom and top 2 deciles start-stops between ahving 0 or 1 start stops
bot_ss <- ss %>% filter(decile<30) #sum(bot_ss$`n()`) == 252 genes in bottom 2 deciles have at least 1 start-stops
top_ss <-  ss %>% filter(decile>80) #sum(top_ss$`n()`) 125 genes in bottom 2 deciles have at least 1 start-stops

#total of genes in bot and top 2 deciles
tot_bot <- sum(bot_ss$total)#3653
tot_top <- sum(top_ss$total)#2864

#total genes with ss in bot and top
botss <- sum(bot_ss$`n()`)
topss<- sum(top_ss$`n()`)

#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene <- c(tot_bot, tot_top)
#how many genes in those deciles with 1+ ss
one <- c(botss, topss)

ss_matrix <- data.frame(zero="",ss=one, total=tot_dec_gene)
ss_matrix$zero <- ss_matrix$total - ss_matrix$ss
#get percentages so can report in paper
ss_matrix$perc_z <- (ss_matrix$zero / ss_matrix$total) *100
ss_matrix$perc_o <- (ss_matrix$ss / ss_matrix$total) *100
#do chi-square
test <- subset(ss_matrix, select = c(zero, ss))
chisq.test(test)$p.value
#p=8.506704e-05
#sig

```
#4. uORF analysis (length and cds-is) 
```{r message=FALSE, error=FALSE, warning=FALSE}
######4.1 uORF length
#bring in data
uo_len <- read.table("uo_ss.txt")
uo_len <- merge(uo_len, length_exon, by="transcript")

uorflen <- merge(gn_final1, uo_len, by="gene_id")
uorflen$decile <- uorflen$decile*10

mean_uole <- mean(uo_len$uorf_len)
ggplot(uorflen, aes(x=factor(decile), y=uorf_len)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("uORF Length (bp)") + geom_hline(yintercept = mean_uole, linetype = "dotted", size=0.8)+ theme(text = element_text(size = 20)) + coord_cartesian(ylim=c(0, 600))


#stats
bot_len <- uorflen %>% filter(decile<30)
#mean
mean(bot_len$uorf_len)#52.46849
top_len <- uorflen %>% filter(decile>80)
mean(top_len$uorf_len)
#59.13161
botl <- bot_len$uorf_len
topl <- top_len$uorf_len

wilcox.test(botl, topl)$p.value

#p=4.315421e-06



######4.2 uORF to CDS distance
#last uORF-CDS distance
#bring in data
uORF_laststop <- read.table("uorfs reinitiation dis.txt")

uORF_laststop <- merge(uORF_laststop, length_exon, by="transcript")

uorf_only <- uORF_laststop %>% filter(orf_type=="uORF")#6320
#merge with gn_final1
gn_dis <- merge(gn_final1, uorf_only, by="gene_id")

#plot dis by decile
gn_dis$decile <- gn_dis$decile*10

mean_dis <- mean(uorf_only$dis)

ggplot(gn_dis, aes(x=factor(decile), y=dis)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("uORF to CDS distance (bp)")  + geom_hline(yintercept = mean_dis,linetype = "dotted", size=0.8) + theme(text = element_text(size = 20))  + coord_cartesian(ylim=c(0, 750))


##stats
bot_dec2 <- gn_dis %>% filter(decile<30)
mean(bot_dec2$dis) #98.83082
top_dec2 <- gn_dis %>% filter(decile>80)
mean(top_dec2$dis) # 76.87811

bot2 <- bot_dec2$dis
top2 <- top_dec2$dis
wilcox.test(bot2, top2)$p.value
#p= 0.0001266438
```
#5. uAUG/length metric
```{r message=FALSE, error=FALSE, warning=FALSE}
atg_length <- read.table("uaug_metric_m1.txt")

#atg_length

names(atg_length)[5]<- "gene_id"
gn_metric <- merge(gnomad_len, atg_length, by = "gene_id")#7375

#plot metric range in boxplots
gn_metric$decile <- gn_metric$decile *10

#add mean line
mean_met <- mean(atg_length$metric)# 0.01040724

ggplot(gn_metric, aes(x=factor(decile), y=metric)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8)  + theme_light() + xlab("LOEUF Decile") + ylab("uAUG/length") + geom_hline(yintercept = mean_met,linetype = "dotted") + theme(text = element_text(size = 20))



#####stats
#top and bottom 2 deciles
top_met <- gn_metric %>% filter(decile>80)
bot_met<- gn_metric %>% filter(decile<30)

#means
mean(top_met$metric) #0.01292813 == 0.013
mean(bot_met$metric) #0.00845449 == 0.009

topmet <- top_met$metric
botmet <- bot_met$metric

wilcox.test(botmet, topmet,)$p.value
#2.774926e-58

```
#6. Ribo-seq uORFs 
```{r message=FALSE, error=FALSE, warning=FALSE}
###1. diff between 0, 1 and 1+ ribo-seq uorfs across deciles
###2. codon optimality of riboseq uorfs across deciles
###3. ribo-seq uorfs start codons across deciles
###4. ribo-seq uorf lengths across deciles


###1. diff between 0, 1 and 1+ ribo-seq uorfs across deciles
#bring in data
smorf_gene <- read.table("smorf_gene.txt")

sm_gn <- merge(gn_final1, smorf_gene, by="gene_id")#3513 

#
no_smorf <- anti_join(gn_final1, smorf_gene, by = "gene_id")#15642
one_smorf <- sm_gn %>% filter(number_uORF == 1) #2574
more_smorf <- sm_gn %>% filter(number_uORF > 1) #939

#group each into deciles
no_smorf_tab <- data.frame(table(no_smorf$decile))
one_smorf_tab <- data.frame(table(one_smorf$decile))
more_smorf_tab <- data.frame(table(more_smorf$decile))

#total in each decile
total_per_decile <- data.frame(table(gn_final1$decile))

no_smorf_tab$total <- total_per_decile$Freq
one_smorf_tab$total <- total_per_decile$Freq
more_smorf_tab$total <- total_per_decile$Freq

#add group name
no_smorf_tab$group <- "0"
one_smorf_tab$group <- "1"
more_smorf_tab$group <- ">1"

#percentage
no_smorf_tab$perc <- no_smorf_tab$Freq/no_smorf_tab$total *100
one_smorf_tab$perc <- one_smorf_tab$Freq/one_smorf_tab$total *100
more_smorf_tab$perc <- more_smorf_tab$Freq/more_smorf_tab$total *100


#rbind to plot
smorf_plot <- rbind(no_smorf_tab, one_smorf_tab, more_smorf_tab)
#reorder bars
smorf_plot$group <- factor(smorf_plot$group, levels = c("0", "1", ">1"))

#change deciles to be multiples of 10
smorf_plot$Var1 <- as.numeric(smorf_plot$Var1)
smorf_plot$Var1 <- smorf_plot$Var1 *10

#plot
ggplot(smorf_plot, aes(x = factor(Var1), y = perc, fill = group )) + geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage") + guides(fill=guide_legend(title="Ribo-seq uORF Number")) + scale_fill_manual(values = c("0" = "#BFACC8", "1" = "#2B4595", ">1" = "#228C80"))+ theme(text = element_text(size = 20))  +# theme(legend.position="none")


#stats
#difference between 0 and 1+ uorfs in bot vs top 2 deciles

###stats for bottom and top 2 deciles start-stops between ahving 0 or 1 uorfs
bot_smrf <-smorf_plot %>% filter(Var1<30) 
top_smrf <-  smorf_plot %>% filter(Var1>80) 

#total of genes in bot and top 2 deciles

#group them as 0 or 1 uorf
bot_smrf[bot_smrf==">1"]<-1
#sum all those in bttom deciles

bot1 <- bot_smrf %>% group_by(group) %>% summarise(Freq=sum(Freq))

top_smrf[top_smrf==">1"]<-1
top1 <- top_smrf %>% group_by(group) %>% summarise(Freq=sum(Freq))

statsuo <- merge(bot1, top1, by="group")
statsuo <- subset(statsuo, select=-c(group))

chisq.test(statsuo)$p.value
#1.692544e-141

#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene_u <- c(tot_bot_u, tot_top_u)
#how many genes in those deciles with 1+ uo
one_uo <- c(botuo, topuo)

uo_matrix <- data.frame(zero="",uo=one_uo, total=tot_dec_gene_u)
uo_matrix$zero <- uo_matrix$total - uo_matrix$uo
#get percentages so can report in paper
uo_matrix$perc_z <- (uo_matrix$zero / uo_matrix$total) *100
uo_matrix$perc_o <- (uo_matrix$uo / uo_matrix$total) *100
#do chi-square
test_uo <- subset(uo_matrix, select = c(zero, uo))
chisq.test(test_uo)$p.value
#p=2.087572e-51
#sig





###2. codon optimality of riboseq uorfs across deciles
#bring in optimality data
ribo_opt2 <- read.table("smorf uorf codon optimality HeLa.txt")
 
gn_rib <- merge(gn_final1, ribo_opt2,by="gene_id")#4834

#plot
gn_rib$decile <- gn_rib$decile *10

mean_co <- mean(ribo_opt2$score)
ggplot(gn_rib, aes(x=factor(decile), y=score)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("Codon Optimality Score") + geom_hline(yintercept = mean_co, linetype = "dotted", size=0.8) + theme(text = element_text(size = 20)) 

#stats
bot_co <- gn_rib %>% filter(decile<30)
top_co <- gn_rib %>% filter(decile>80)

botco <- bot_co$score
topco <- top_co$score
wilcox.test(botco, topco)
#p-value = 0.1688

###3. ribo-seq uorfs start codons across deciles
#from smorf2 data
uORFs <- read.table("ribo_seq_uo.txt")
#merge with gnomad
gn_smorf <- merge(gn_final1, uORFs,by="gene_id")

#i want to plot atg vs not atg across loeuf
gn_smorf$start_type <- "non-canonical"

#if ATG change to "canonical"
for(l in 1:nrow(gn_smorf)){
  if (gn_smorf$starts[l] == "ATG"){
    gn_smorf$start_type[l] <- (gn_smorf$start_type[l] <- "canonical")
  }
}

#aggregate start type by decile
#tab_starts <- aggregate(gn_smorf$start_type, by = list(gn_smorf$decile), FUN=sum)#not working
starts_tab <- gn_smorf %>% group_by(decile, start_type)%>%summarise(n())
starts_tab2 <- gn_smorf %>% group_by(decile)%>%summarise(n()) #how many in each decile
#append total to starts_tab
dec_tot <- starts_tab2$`n()`
dec_tot2 <- rep(dec_tot, each=2)
#get percentages
starts_tab$total <- dec_tot2
starts_tab$perc <- starts_tab$`n()`/starts_tab$total *100

#plot
starts_tab$decile <- starts_tab$decile*10
starts_tab$start_type <- factor(starts_tab$start_type, levels=c("canonical","non-canonical"))

ggplot(starts_tab, aes(x=factor(decile), y=perc, fill=start_type)) + geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage")  + guides(fill=guide_legend(title="Start Type")) + scale_fill_manual(values = c("canonical" = "#BFACC8", "non-canonical" = "#2B4595")) + theme(text = element_text(size = 20))  + theme(legend.title=element_blank())  #+ theme(legend.position="none")


#in black and grey
ggplot(starts_tab, aes(x=factor(decile), y=perc, fill=start_type)) + geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage")  + guides(fill=guide_legend(title="Start Type")) + scale_fill_manual(values = c("canonical" = "grey", "non-canonical" = "black")) + theme(text = element_text(size = 20))  + theme(legend.title=element_blank())  #+ theme(legend.position="none")

#export as table
names(starts_tab)[3]<-"count"
names(starts_tab)[4]<-"total in decile"
names(starts_tab)[5]<-"percentage"

#write_xlsx(starts_tab , "2F_riboseq_starts.xlsx")

#stats
names(starts_tab)[3]<-"n"

bot_start <- starts_tab %>% filter(decile<30)
top_start <- starts_tab %>% filter(decile>80)

#group by and sum 
bot <- bot_start %>% group_by(start_type) %>% summarise(n=sum(n))
top <- top_start %>% group_by(start_type) %>% summarise(n=sum(n))
names(bot)[2]<-"bot"
names(top)[2]<-"top"

can_stats<- merge(bot, top, by="start_type")

#transpose?
cs2 <- as.data.frame(t(can_stats)) 
#remove top line
cs2 <- cs2[-1,]

cs2$V1 <- as.numeric(cs2$V1)
cs2$V2 <- as.numeric(cs2$V2)
chisq.test(cs2)$p.value
#0.183124





###4. ribo-seq uorf lengths across deciles
#gn_smorf has length in it (col:len)
gn_smorf$decile <- gn_smorf$decile*10

mean_rs <- mean(uORFs$len)
ggplot(gn_smorf, aes(x=factor(decile), y=len)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("Ribo-seq uORF Length (bp)") + geom_hline(yintercept = mean_rs, linetype = "dotted", size=0.8) + theme(text = element_text(size = 20)) + coord_cartesian(ylim=c(0, 600))

#stats
bot_rs <- gn_smorf %>% filter(decile<30)
top_rs <- gn_smorf %>% filter(decile>80)

botrs <- bot_rs$len
toprs <- top_rs$len

wilcox.test(botrs, toprs)
#p=0.9453
```
#7. Translational Efficiency (Noderer) 
```{r message=FALSE, error=FALSE, warning=FALSE}
#from TE for uAUG markdown - ORF_TE df
#want boxplot of TE for each disease set
#bring in TE data
ORF_te2 <- read.table("orf_TE_noderer_m1.txt")
#think this is called ORF_te2 now
#names(ORF_te2)[5]<- "transcript"
te <- merge(ORF_te2, length_exon, by="transcript")


TE_gnomad <- merge(gn_final1, te, by="gene_id")

#just uorfs
te_uo <- TE_gnomad %>% filter(orf_type=="uORF")
mean_te <- mean(te_uo$efficiency)#78.77569

te_uo$decile <- te_uo$decile*10

ggplot(te_uo, aes(x=factor(decile), y=efficiency)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("TE (Norderer)") + geom_hline(yintercept = mean_te,linetype = "dotted", size=0.8) + theme(text = element_text(size = 20))

#stats
bot_te <- te_uo %>% filter(decile<30)
top_te <- te_uo %>% filter(decile>80)

bote <- bot_te$efficiency
tope <- top_te$efficiency

wilcox.test(bote, tope)$p.value
# 0.583095

```
#8. PhyloP 
```{r message=FALSE, error=FALSE, warning=FALSE}
#1. bring in PhyloP data
####5'UTR mean PhyloP
phylop <- read.table("five_utr_ph_m1.txt")
phylop <- cSplit(phylop, "gene_id", ".")
names(phylop)[7]<-"gene_id"
#merge
utr_ph <- merge(phylop, gn_final1,by="gene_id")

####uORF (start and ends) and start-stops
phy2 <- read.table("uorf_ss phylop m1.txt")
#merge with mane to get gene names
names(phy2)[1]<-"transcript"
ph_name <- merge(length_exon, phy2, by="transcript")

#merge gnomad
ph_gn <- merge(gn_final1, ph_name, by="gene_id")

########plot mean 5'UTR, uorf start, uorf end and start stops
#2. combine to make one df
names(utr_ph)[3]<-"phylop"
utr_ph <- subset(utr_ph, select=c(gene.x, gene_id, phylop,decile))
utr_ph$group <- "5'UTR"
utr_ph$decile <- utr_ph$decile*10

#append to ph_gn
ph_gn$decile<-ph_gn$decile*10
#subset
ph_gn2 <- subset(ph_gn, select=c(gene.x, gene_id, phylop,decile, group))

plot_ph <- rbind(ph_gn2, utr_ph)
 #change all uorf end to uorf stop
plot_ph[plot_ph=="uORF end"]<- "uORF Stop"

#plot
#order
plot_ph$group <- factor(plot_ph$group, levels = c("5'UTR", "uORF Start", "uORF Stop", "Start-Stop"))

ggplot(plot_ph, aes(x=factor(decile), y=phylop, fill=group)) + geom_boxplot() + theme_light() + xlab("LOEUF Decile (%)") + ylab("PhyloP") + scale_fill_manual(values = c("5'UTR" = "#793A92", "uORF Start" = "navy", "uORF Stop" = "navy", "Start-Stop" = "deepskyblue1"))  + geom_hline(yintercept =2,linetype = "dotted", size=0.6) + theme(text = element_text(size = 20))  + theme(legend.title=element_blank()) #+ theme(legend.position = "none") 


############STUDENT T TEST
#for 5UTR mean phylop
#1.2 top and bottom 2 deciles
bot2 <- utr_ph %>% filter(decile<30)

x <- sum(bot2$phylop[which(bot2$phylop>=0)])
y <- sum(bot2$phylop[which(bot2$phylop<=0)])
z <- x+y #(x-y added it)
#get number of values in bot
z/nrow(bot2) 
#mean = 1.101596

top2 <- utr_ph %>% filter(decile>80)

x <- sum(top2$phylop[which(top2$phylop>=0)])
y <- sum(top2$phylop[which(top2$phylop<=0)])
z <- x+y #(x-y added it)
#get number of values in bot
z/nrow(top2) 
#0.01249374

#students t-test
t.test(bot2$phylop, top2$phylop)
#p-value < 2.2e-16
#but when i add $p.value = 0 and when i put the test into a variable it says p val = 0



#stats for uorf start, ends and start-stops also needed
bot <- ph_gn2 %>% filter(decile<30)
top <- ph_gn2 %>% filter(decile>80)

###uorf start
bot_uos <- bot %>% filter(group=="uORF Start")
top_uos <- top %>% filter(group=="uORF Start")

botuos <- bot_uos$phylop
topuos <- top_uos$phylop

t.test(botuos,topuos)$p.value
#p-value < 2.2e-16

####uorf end
bot_uoe <- bot %>% filter(group=="uORF end")
top_uoe <- top %>% filter(group=="uORF end")

botuoe <- bot_uoe$phylop
topuoe <- top_uoe$phylop

t.test(botuoe,topuoe)$p.value
#p-value < 2.2e-16

####start-stops
bot_ss <- bot %>% filter(group=="Start-Stop")
top_ss <- top %>% filter(group=="Start-Stop")

botss <- bot_ss$phylop
topss <- top_ss$phylop

t.test(botss,topss)$p.value
#p-value < 2.2e-16


###make tables for export##

five_ph$phylop <- round(five_ph$phylop, digits = 4)#round them
fiveph_tab <- five_ph %>% group_by(decile) %>% summarise(mean_phylop=mean(phylop, na.rm=TRUE), median_phylop=ifelse(length(phylop)>1, median(phylop, na.rm = TRUE), phylop),sd_phylop=sd(phylop, na.rm=TRUE), min_phylop = min(phylop, na.rm = TRUE), max_phylop = max(phylop, na.rm = TRUE),range = paste(min(phylop, na.rm = TRUE), max(phylop, na.rm = TRUE), sep = ":"))
fiveph_tab$group <- "5'UTR"

#write_xlsx(fiveph_tab, "5utrph.xlsx")


#table for uorf starts
start_ph <- plot_ph %>% filter(group=="uORF Start")
start_ph$phylop <- round(start_ph$phylop, digits = 4)#round them
start_ph_tab <- start_ph %>% group_by(decile) %>% summarise(mean_phylop=mean(phylop, na.rm=TRUE), median_phylop=ifelse(length(phylop)>1, median(phylop, na.rm = TRUE), phylop),sd_phylop=sd(phylop, na.rm=TRUE), min_phylop = min(phylop, na.rm = TRUE), max_phylop = max(phylop, na.rm = TRUE),range = paste(min(phylop, na.rm = TRUE), max(phylop, na.rm = TRUE), sep = ":"))
start_ph_tab$group <- "uORF Start"
#remove the range one? can do in excel or come back later if not right?
#write_xlsx(start_ph_tab, "uostart_ph.xlsx")

#table for uorf stops
stop_ph <- plot_ph %>% filter(group=="uORF Stop")
stop_ph$phylop <- round(stop_ph$phylop, digits = 4)#round them
stop_ph_tab <- stop_ph %>% group_by(decile) %>% summarise(mean_phylop=mean(phylop, na.rm=TRUE), median_phylop=ifelse(length(phylop)>1, median(phylop, na.rm = TRUE), phylop),sd_phylop=sd(phylop, na.rm=TRUE), min_phylop = min(phylop, na.rm = TRUE), max_phylop = max(phylop, na.rm = TRUE),range = paste(min(phylop, na.rm = TRUE), max(phylop, na.rm = TRUE), sep = ":"))
stop_ph_tab$group <- "uORF Stop"
#remove the range one? can do in excel or come back later if not right?
#write_xlsx(stop_ph_tab, "uostop_ph.xlsx")

#table for start stops
ss_ph <- plot_ph %>% filter(group=="Start-Stop")
ss_ph$phylop <- round(ss_ph$phylop, digits = 4)#round them
ss_ph_tab <- ss_ph %>% group_by(decile) %>% summarise(mean_phylop=mean(phylop, na.rm=TRUE), median_phylop=ifelse(length(phylop)>1, median(phylop, na.rm = TRUE), phylop),sd_phylop=sd(phylop, na.rm=TRUE), min_phylop = min(phylop, na.rm = TRUE), max_phylop = max(phylop, na.rm = TRUE),range = paste(min(phylop, na.rm = TRUE), max(phylop, na.rm = TRUE), sep = ":"))
ss_ph_tab$group <- "Start-Stop"
#remove the range one? can do in excel or come back later if not right?
#write_xlsx(ss_ph_tab, "ss_ph.xlsx")

##### re-run phylop analysis on 100-300bp long 5'utr #####
#go to this section and then change this to keep utr length

#2. combine to make one df
names(utr_ph)[3]<-"phylop"
utr_ph <- subset(utr_ph, select=c(gene.x, gene_id, phylop,decile,UTR_length))
utr_ph$group <- "5'UTR"
utr_ph$decile <- utr_ph$decile*10

#append to ph_gn
ph_gn$decile<-ph_gn$decile*10
#subset
ph_gn2 <- subset(ph_gn, select=c(gene, gene_id, phylop,decile, group,UTR_length))
names(utr_ph)[1]<-"gene"
plot_ph <- rbind(ph_gn2, utr_ph)
 #change all uorf end to uorf stop
plot_ph[plot_ph=="uORF end"]<- "uORF Stop"

#restrict to 1000-300bps
mid_ph <- plot_ph %>%  filter(UTR_length %in% 100:300)
#save
#write.table(mid_ph, "ph_shorterutrs.txt")


#t-test bottom vs top 2 deciles for each category - i think 5'utr length only
short_utr <- mid_ph %>% filter(group=="5'UTR")

bot_ph <- short_utr %>% filter(decile<30)
top_ph <- short_utr %>% filter(decile>80)


botph<- bot_ph$phylop
topph <- top_ph$phylop

t.test(botph, topph, na.rm=TRUE)$p.value
#p=1.207976e-232
```
#9.GC content
```{r message=FALSE, error=FALSE, warning=FALSE}
gc_save<- read.table("man1_gc.txt")

gc_save <- merge(length_exon, gc_save, by="transcript")

gc_gnomad <- merge(gn_final1, gc_save, by="gene_id")

#plot
gc_gnomad$decile <- gc_gnomad$decile*10

#add mean intercept
mean_gc <- mean(gc_save$gc_percent)

ggplot(gc_gnomad, aes(x=factor(decile), y=gc_percent)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("GC Content per bp (%)") + geom_hline(yintercept = mean_gc,linetype = "dotted", size=0.8) + theme(text = element_text(size = 20))


###stats
#chi square on bottom and top 2 deciles GC content

bot_gc <- gc_gnomad %>% filter(decile<30) 
mean(bot_gc$gc_percent)#67.3%
top_gc <-  gc_gnomad %>% filter(decile>80)
mean(top_gc$gc_percent)#60%

bot_gc <- bot_gc$gc_percent
top_gc <- top_gc$gc_percent

wilcox.test(bot_gc, top_gc)$p.value
#4.946138e-134

#do stats on when u restrict it to 100-300bps
mid<- gc_gnomad %>% filter(utr_len %in% 100:300) #6514

bot_gc1 <- mid %>% filter(decile<30) 
mean(bot_gc1$gc_percent)#69%
top_gc1 <-  mid %>% filter(decile>80)
mean(top_gc1$gc_percent)#60%

bot_gc1 <- bot_gc1$gc_percent
top_gc1 <- top_gc1$gc_percent

wilcox.test(bot_gc1, top_gc1)$p.value
# 2.350705e-89
```
#10.CAGE
```{r message=FALSE, error=FALSE, warning=FALSE}
cage_count <- read.table("cage m1.txt")
cage_count <- cSplit(cage_count, "gene_id", ".")
names(cage_count)[6]<-"gene_id"
cage_gn <- merge(gn_final1, cage_count, by="gene_id")#16233

#add groupings for how i want to plot..
#gna do 1, 2-5, 6+

cage_one <- cage_gn %>% filter(count <2)#2528
cage_mid <- cage_gn%>% filter(count >= 2 & count <= 5)#8536
cage_six <- cage_gn %>% filter(count >5)#4807

cage_one$group <- "1"
cage_mid$group <- "2-5"
cage_six$group <- "≥6"

cage_pl <- rbind(cage_one, cage_mid, cage_six)
cage_pl2 <- cage_pl %>% group_by(decile, group) %>% summarise(n())

#add total col
total <- data.frame(table(cage_gn$decile))
totals2 <- rep(total$Freq, each=3)
cage_pl2$total <- totals2

#percentage
cage_pl2$perc <- cage_pl2$`n()`/cage_pl2$total *100
cage_pl2$decile <-cage_pl2$decile*10

#plot
cage_pl2$group <- factor(cage_pl2$group, levels=c("1", "2-5", "≥6") )

#mean cage peak
#3 shades of green
ggplot(cage_pl2, aes(x=factor(decile), y=perc, fill=group))+ geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage")+ guides(fill=guide_legend(title="CAGE Peaks")) + scale_fill_manual(values = c("1" = "#7CDFD4", "2-5" = "#228C80", "≥6" = "#14524B")) + theme(text = element_text(size = 20)) #+ theme(legend.position="none") 

#+ theme(text = element_text(size = 15)) for leg

#its not coming up the more than sign when exporting plot so manially added it in ppt



###stats###
#showing that more peaks in lowest 2 dec vs highest
#diff between 1 and 1+ peaks?


bot_cage <- cage_gn %>% filter(decile<3)
top_cage <- cage_gn %>% filter(decile>8)


#bot 2 deciles
bot_cage <-bot_cage%>% mutate(one_cutoff = count>1)#differentiate above and below 1
bot_cage <- bot_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "TRUE", "1+"))#change to meaningful vals
bot_cage <- bot_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "FALSE", "1"))
total <- nrow(bot_cage) 
b <- data.frame(table(bot_cage$one_cutoff))
b$total <- total
b$perc <- (b$Freq/b$total)*100
#transpose
b3 <- as.data.frame(t(b))
#select the only row I want
b4<- b3 %>% slice(2)
names(b4)[1]<-"1"
names(b4)[2]<-"1+"
rownames(b4) <- c("bot")
b4$`1` <- as.numeric(b4$`1`)
b4$`1+` <- as.numeric(b4$`1+`)
#round
b4<- b4 %>% mutate_if(is.numeric, round)

#top 2 deciles

top_cage <-top_cage%>% mutate(one_cutoff = count>1)#differentiate above and below 1
top_cage <- top_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "TRUE", "1+"))#change to meaningful vals
top_cage <- top_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "FALSE", "1"))
total2 <- nrow(top_cage) 
t <- data.frame(table(top_cage$one_cutoff))
t$total <- total2
t$perc <- (t$Freq/t$total)*100
#transpose
df3 <- as.data.frame(t(t))
#select the only row I want
df4<- df3 %>% slice(2)
names(df4)[1]<-"1"
names(df4)[2]<-"1+"
rownames(df4) <- c("top")
df4$`1` <- as.numeric(df4$`1`)
df4$`1+` <- as.numeric(df4$`1+`)
#round
df4<- df4 %>% mutate_if(is.numeric, round)

#bind
cage_stat <- rbind(b4, df4)

chisq.test(cage_stat)$p.value
#1.518424e-89

###stats for > or equal 6 peaks
bot_cage <-bot_cage%>% mutate(one_cutoff = count>=6)#differentiate above and below 1
bot_cage <- bot_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "TRUE", "6+"))#change to meaningful vals
bot_cage <- bot_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "FALSE", "<6"))
total <- nrow(bot_cage) 
b <- data.frame(table(bot_cage$one_cutoff))
b$total <- total
b$perc <- (b$Freq/b$total)*100
#transpose
b3 <- as.data.frame(t(b))
#select the only row I want
b4<- b3 %>% slice(2)
names(b4)[1]<-"<6"
names(b4)[2]<-"6+"
rownames(b4) <- c("bot")
b4$`<6` <- as.numeric(b4$`<6`)
b4$`6+` <- as.numeric(b4$`6+`)
#round
b4<- b4 %>% mutate_if(is.numeric, round)

#top 2 deciles

top_cage <-top_cage%>% mutate(one_cutoff = count>=6)#differentiate above and below 1
top_cage <- top_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "TRUE", "6+"))#change to meaningful vals
top_cage <- top_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "FALSE", "<6"))
total2 <- nrow(top_cage) 
t <- data.frame(table(top_cage$one_cutoff))
t$total <- total2
t$perc <- (t$Freq/t$total)*100
#transpose
df3 <- as.data.frame(t(t))
#select the only row I want
df4<- df3 %>% slice(2)
names(df4)[1]<-"<6"
names(df4)[2]<-"6+"
rownames(df4) <- c("top")
df4$`<6` <- as.numeric(df4$`<6`)
df4$`6+` <- as.numeric(df4$`6+`)
#round
df4<- df4 %>% mutate_if(is.numeric, round)

#bind
cage_stat <- rbind(b4, df4)

chisq.test(cage_stat)$p.value
#8.059105e-111


#tables of data for export
#cage_pl2 is the right thing, just reorder

cage_2 <- cage_pl2 %>% arrange(group, match(group, c("1", "2-5", ">6")))
names(cage_2)[2]<-"cage group"
names(cage_2)[3]<-"count"
names(cage_2)[4]<-"total in decile"
names(cage_2)[5]<-"percentage"
#write_xlsx(cage_2 , "2E_gn_cage.xlsx")
```
#11. MFE
```{r message=FALSE, error=FALSE, warning=FALSE}
energy_score <- read.table("freefold energyscore m1.txt")

energy_score <- merge(length_exon, energy_score, by="transcript")

#merge with gnomad, boxplot across deciles
gn_energy <- merge(gn_final1, energy_score, by="gene_id")

#plot
gn_energy$decile <- gn_energy$decile*10


#plot the raw scores too
#add mean energy for all genes
mean(energy_score$energy) #-78.81686


#try raincloud plot
ggplot(data = gn_energy, aes(x=factor(decile), y = energy, fill = decile)) +   geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8, fill="#793A92") +     geom_point(aes(y = energy, fill = decile), position = position_jitter(width = 0.15), size = 1, alpha = 0.1, color="grey48") +     geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.5, fill="#793A92") +     labs(x = "LEOUF Decile", y = "Minimum Folding Energy") + guides(fill = FALSE, color = FALSE) +theme_light() + theme(text = element_text(size = 20)) + coord_cartesian(ylim=c(0, -1000))

####stats
bot_mfe <- gn_energy %>% filter(decile<30)
mean(bot_mfe$energy)#-115.1429
top_mfe <- gn_energy %>% filter(decile>80)
mean(top_mfe$energy)#-54.80042

#wilcoxon
botmfe <- bot_mfe$energy
topmfe <- top_mfe$energy
wilcox.test(botmfe, topmfe)$p.value
#7.349057e-206


#do stats on when u restrict it to 100-300bps
mid<- gn_energy %>% filter(utr_len %in% 100:300) #6514

bot2 <- mid %>% filter(decile<30)
mean(bot2$energy)#-81.52125
top2 <-mid %>% filter(decile>80)
mean(top2$energy)#-60.01709

#wilcoxon
bot_2 <- bot2$energy
top_2 <- top2$energy
wilcox.test(bot_2, top_2)$p.value
# 2.976304e-57

##table of plotted data for supp fig for export
gnenergy_tab <- gn_energy %>% group_by(decile) %>% summarise(mean_mfe=mean(energy), median_mfe=median(energy),range_mfe=paste(min(energy),max(energy), sep="-"), sd_mfe=sd(energy))

#write_xlsx(gnenergy_tab, "2c_mfe.xlsx")
```
#12. CADD 
```{r message=FALSE, error=FALSE, warning=FALSE}
#bring in CADD data
cadd <- read.table("cadd_all_m1.txt")
names(cadd)[1] <-"transcript"
cadd2 <- merge(cadd, length_exon, by="transcript")


#merge with gnomad
gn_cadd <- merge(gn_final1, cadd2, by="gene_id")

#plot
gn_cadd$decile <- gn_cadd$decile*10


gn_cadd$group <- factor(gn_cadd$group, levels = c("5'UTR", "uORF Start", "uORF Stop", "Start-Stop"))


ggplot(gn_cadd, aes(x=factor(decile), y=cadd, fill=group)) + geom_boxplot() + theme_light() + xlab("LOEUF Decile") + ylab("CADD") + scale_fill_manual(values = c("5'UTR" = "#793A92", "uORF Start" = "navy", "uORF Stop" = "navy", "Start-Stop" = "deepskyblue1")) + theme(text = element_text(size = 20)) # + theme(legend.position="none")  
#stats
#mean of 5'UTR CADD bottom and top deciles
cad_utr <- gn_cadd %>% filter(group=="5'UTR")
#function
means <- function(df){
  x <- sum(df$cadd[which(df$cadd>=0)])
y <- sum(df$cadd[which(df$cadd<=0)])
z <- x+y #(x-y added it)
#get number of values in df
return(z/nrow(df))
}

#comparing top and bottom 2 deciles
bot2 <- cad_utr %>% filter(decile<30)
means(bot2)
#1.483858
top2<- cad_utr %>% filter(decile>80)
means(top2)
#0.6942674

#stats
#5utr cadd
t.test(bot2$cadd, top2$cadd)
#p-value < 2.2e-16

#uo start
cad_uosta <- gn_cadd %>% filter(group=="uORF Start")
bot_u <- cad_uosta %>% filter(decile<30)
top_u <-cad_uosta %>% filter(decile>80)

t.test(bot_u$cadd, top_u$cadd)
#p-value < 2.2e-16

#uo stop
cad_uostp <- gn_cadd %>% filter(group=="uORF Stop")
bot_u2 <- cad_uostp %>% filter(decile<30)
top_u2 <-cad_uostp %>% filter(decile>80)

t.test(bot_u2$cadd, top_u2$cadd)
#p-value < 2.2e-16

#ss
cad_ss <- gn_cadd %>% filter(group=="Start-Stop")
bot_ss <- cad_ss %>% filter(decile<30)
top_ss <-cad_ss %>% filter(decile>80)

t.test(bot_ss$cadd, top_ss$cadd)
#p-value < 2.2e-16


####redo cadd 5'utr scores on 100-300bps, bottom vs top 2 leouf deciles
mid_cadd <- gn_cadd %>% filter(UTR_length %in% 100:300)
mid_cadd_utr <- mid_cadd %>% filter(group=="5'UTR")

botcad <- mid_cadd_utr %>% filter(decile<30)
topcad <- mid_cadd_utr %>% filter(decile>80)
botcad2 <- botcad$cadd
topcad2 <- topcad$cadd

t.test(botcad2, topcad2,na.rm=TRUE)$p.value
#still giving 0


t.test(botcad$cadd, topcad$cadd)
```


#13. uAUG Consensus sequences
```{r}
#compare the uAUG consensus sequences for top and bottom 2 LEOUF deciles
#bring in all uaugs
uo_ss <- read.table("uo_ss.txt")
oorf <- read.table("oorf.txt")

#put into 1 table
oorf$orf_type <- "oORF"

uoss2 <- subset(uo_ss, select=-c(stop_pos))
oorf2 <- subset(oorf, select=-c(utr_len, oorfend, three, integer, frame, oORF_seq))
aug <- rbind(uoss2, oorf2)

#get range of sequence around aug (excludes those v close to CDS and utr start)
minus <- aug %>% filter(atg_pos>15)
#get+15
minus$atg_end <- minus$atg_pos+2
minus$utr_len <- str_length(minus$sequence)
minus$end_dis <- minus$utr_len - minus$atg_end
plus <- minus %>% filter(end_dis>15)
plus$minus <- str_sub(plus$sequence, start = plus$atg_pos-15, end =plus$atg_pos-1) 
plus$plus <-  str_sub(plus$sequence, start=plus$atg_end+1, end= plus$atg_end+15)
plus$plot <- paste(plus$minus, plus$plus, sep="XXX")

require(ggplot2)
require(ggseqlogo)

#merge with gnomad
names(gnomad_len)[3]<-"transcript"
gn_koz <- merge(gnomad_len, plus, by="transcript")

bot_koz <- gn_koz %>% filter(decile<30)
top_koz <- gn_koz %>% filter(decile>80)

#plot bottom 2 deciles
ggplot() + geom_logo(bot_koz$plot) + theme_logo() +ggtitle("bottom louef uaugs") + ylab("Information Content")
#plot top 2 deciles
ggplot() + geom_logo(top_koz$plot) + theme_logo() +ggtitle("top louef uaugs") + ylab("Information Content")




################ kozak consensus for uORFs across loeuf################

#calculate kozak scores for each uorf
#plot by decile
#bring in uORFs
uo_len <- read.table("uo_ 20.9.22.txt")


#filter out any atgs that are too close to TSS or CDS i want the -3 and +1 pos
uo_len$utr_len <- str_count(uo_len$sequence)  

#if too close to TSS
uo1 <- uo_len %>% filter(!atg_pos<4)
#if too close to CDS..
uo1$cds_dis <- uo1$utr_len-uo1$atg_pos
#stupid.. they are uorfs so wont be too close to cds duh

#-3 pos:
uo1$prekoz <- substr(uo1$sequence, start=(uo1$atg_pos-3), stop=(uo1$atg_pos-3))
#+1 pos
uo1$postkoz <- substr(uo1$sequence, start=(uo1$atg_pos+3), stop=(uo1$atg_pos+3))

#set up count column for the kozak important bases
uo1$pre_g <- 0
uo1$pre_a <- 0
uo1$post_g <- 0

#loop to +1 if any of these are present
# if pre = G
for(t in 1:nrow(uo1)){
  if (uo1$prekoz[t] == "G"){
    uo1$pre_g[t] <- (uo1$pre_g[t] + 1)
  }
}
#if pre = A
for(tt in 1:nrow(uo1)){
  if (uo1$prekoz[tt] == "A"){
    uo1$pre_a[tt] <- (uo1$pre_a[tt] + 1)
  }
}
#if post = G
for(ttt in 1:nrow(uo1)){
  if (uo1$postkoz[ttt] == "G"){
    uo1$post_g[ttt] <- (uo1$post_g[ttt] + 1)
  }
}

#determine strengths
#add either a or g pre plus post g to get a total
uo1$pre_apluspostg <- uo1$pre_a + uo1$post_g
uo1$pre_gpluspostg <- uo1$pre_g + uo1$post_g

#add these to get an overall total
uo1$totalkoz <- uo1$pre_apluspostg + uo1$pre_gpluspostg
#going to be 0, 1, 2, or 3
#pre A/G + post G = strong (3)
#pre A/G = moderate (1)
#post G = moderate (2)
#none = weak (0)

#write nice loop here to add a strength
#turn into dt
uo1 <- as.data.table(uo1)
uo1[totalkoz=="3", kozak := "STRONG"]
uo1[totalkoz=="2", kozak := "MODERATE"]
uo1[totalkoz=="1", kozak := "MODERATE"]
uo1[totalkoz=="0", kozak := "WEAK"]

#now merge with gnomad
#need gene id
names(uo1)[1]<-"transcript_id"
orf2 <- merge(uo1, length_exon, by="transcript_id")#17827 uorfs


#merge with gnomad
#cut off after full stop
orf2 <- cSplit(orf2, "gene_id", ".") 
names(orf2)[24]<-"gene_id"
koz_gn <- merge(gn_final1, orf2, by="gene_id")#15879

#get kozak strengths per decile
#get total per decile - total uorfs per decile
total <- data.frame(table(koz_gn$decile))
totals <- total$Freq
totals2 <- rep(totals, each=3)
#aggregate to get total of each kozak type per decile
koz2 <- koz_gn %>% group_by(decile,kozak) %>% summarise(n()) 
koz2$total <- totals2
koz2$perc <- koz2$`n()`/koz2$total *100

koz2$kozak <- factor(koz2$kozak, levels = c("MODERATE","WEAK" , "STRONG"))
koz2$decile <-koz2$decile*10 

ggplot(koz2, aes(x=factor(decile), y=perc, fill=kozak)) + geom_bar(position="dodge", stat="identity") + theme_light() + xlab("LOEUF Decile") + ylab("Percentage") + scale_fill_manual(values = c("WEAK" = "#BFACC8", "MODERATE" = "#2B4595", "STRONG" = "#228C80"))  + theme(text = element_text(size = 20))

#+ theme(legend.title=element_blank()) + theme(text = element_text(size = 20)) #+ theme(legend.position = "none")

#do stats - just on strong ones?


###stats for bottom and top 2 deciles start-stops between ahving 0 or 1 uorfs
bot_koz <-koz2 %>% filter(decile<30) 
top_koz <-  koz2 %>% filter(decile>80) 

#total of genes in bot and top 2 deciles
tot_bot_u <- 2278+2094
tot_top_u <- 1433+1161

#total genes with strong kozak in bot and top
botk <- bot_koz %>% filter(kozak=="STRONG")
topk <- top_koz %>% filter(kozak=="STRONG")

botk <- sum(botk$`n()`)
topk <- sum(topk$`n()`)


#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene_u <- c(tot_bot_u, tot_top_u)
#how many genes in those deciles with strong
one_uo <- c(botk, topk)

uo_matrix <- data.frame(zero="",uo=one_uo, total=tot_dec_gene_u)
uo_matrix$zero <- uo_matrix$total - uo_matrix$uo
#get percentages so can report in paper
uo_matrix$perc_z <- (uo_matrix$zero / uo_matrix$total) *100
uo_matrix$perc_o <- (uo_matrix$uo / uo_matrix$total) *100
#do chi-square
test_uo <- subset(uo_matrix, select = c(zero, uo))
chisq.test(test_uo)$p.value
#p=9.033734e-05
#sig

#this is saying that the difference between proportions of genes with strong kozaks is sig between top and bottom declies

```



#14. CDS length across LOEUF
```{r}
select_ensembl_genomic <- read.delim("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE downloads/MANE v1/MANE.GRCh38.v1.0.ensembl_genomic.gff", header=FALSE)
ensembl_genomic <- select_ensembl_genomic
cds <- subset(ensembl_genomic, ensembl_genomic$V3 == "CDS")

cds <- cSplit(cds, "V9", ";")
cds$V9_03 <- str_remove_all(cds$V9_03, "[gene_id=]") 
cds$V9_04 <- str_remove_all(cds$V9_04, "[transcript_id=]")
cds$V9_06 <- str_remove_all(cds$V9_06,  "[gene_name=]") 
cds$V9_08 <- str_remove_all(cds$V9_08, "[transcript_name=]")
cds$V9_09 <- str_remove_all(cds$V9_09, "[exon_number=]")
cds$V9_11 <- str_remove_all(cds$V9_11, "[tag=]")
names(cds)[11] <- "gene_id"
names(cds)[12] <- "transcript_id"
names(cds)[14] <- "gene_name"
names(cds)[16] <- "transcript_name"
names(cds)[17] <- "exon_number"
names(cds)[19] <- "MANE_tag"


cds <-  cds[grepl("MANE_Selec", cds$MANE_tag), ]
#length
cds$len <- (cds$V5 - cds$V4) +1

#aggreagte to get cds exon length per transcript

cds_len <- cds %>% group_by(gene_id) %>% summarise(len=sum(len))
#save
#write.table(cds_len, "/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/cds_len.txt")

cds_len <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/cds_len.txt")

#merge with gnomad and plot cds lengths by decile
cds_len <- cSplit(cds_len, "gene_id", ".")
names(cds_len)[2]<-"gene_id"
#names(gn_final1)[10]<-"gene_id"
gn_cds <- merge(gn_final1, cds_len, by="gene_id")


gn_cds$decile <- gn_cds$decile*10

ggplot(gn_cds, aes(x=factor(decile), y=len)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("CDS Length (bp)") + ylim(0,30000)  + theme(text = element_text(size = 20))
#in dec 2 ther eis one outlier >90k

##also plot a denisty plot of CDS lengths across all genes
ggplot(data = cds_len, mapping = aes(x = len)) + geom_density(color="black", fill="lightgrey") + theme_classic() + xlab("CDS Length (bp)") + ylab("Density")

####thesis####
#some stats to show that lower leouf deciles have longer CDS? 

bot <- gn_cds %>% filter(decile<30)
mean(bot$len)#2720.293

top <-gn_cds %>% filter(decile>80)
mean(top$len)#832.7367
bot2 <- bot$len
top2 <- top$len
wilcox.test(bot2, top2)$p.value
#giving 0. 
#try remove TTN

bot3 <- bot %>% filter(!gene=="TTN")
bot2 <- bot3$len
top2 <- top$len
wilcox.test(bot2, top2)$p.value
#doesnt help.
#not sure what to do here

#try good ol correlation thing?
#remove ttn
gn_cds_nottn <- gn_cds %>% filter(!gene=="TTN")
ggplot(gn_cds_nottn, aes(decile, len)) +
     geom_point(alpha = 0.4) +
     geom_smooth(method = 'lm', color = "#793A92") + 
     annotate("text", x = Inf, y = Inf, hjust = 1, vjust = 1,
              label = paste("R² = ", round(summary(lm(len ~ decile, data = gn_cds_nottn))$r.squared, 3)),
              color = "#793A92") + theme_light()

#+ xlab("5'UTR Length (bp)") + ylab("uAUG Count") + theme(text = element_text(size = 20))
#r2=0.158

#just leave out




#take shortest 5% genes (shortest CDS) out , rebin in to deciles and replot the 5'UTR length plot across LEOUF and see pattern
#first do on the ones already merged with gnomad

# Calculate the 5th percentile
threshold <- quantile(gn_cds$len, probs = 0.05)

# Subset the dataframe to remove genes with length <= threshold
genes_subset <- subset(gn_cds, len > threshold)

#rebin
genes_subset$decile <- ntile(genes_subset$oe_lof_upper, 10)#16614

#plot their 5utr length across deciles now
#not all of them have 5utrs..?
five_len <- subset(length_exon, select=c(gene_id,UTR_length))

#merge
gn_five <- merge(five_len, genes_subset, by="gene_id")#16456

#plot 5utr length
ggplot(gn_five, aes(x=factor(decile), y=UTR_length)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("UTR Length (bp)") 

#replot cds length across deciles with shortest genes remvoed
ggplot(gn_five, aes(x=factor(decile), y=len)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()


##bottom 10% removed
#take shortest 10% genes (shortest CDS) out , rebin in to deciles and replot the 5'UTR length plot across LEOUF and see pattern
#first do on the ones already merged with gnomad

# Calculate the 10th percentile
threshold <- quantile(gn_cds$len, probs = 0.10)

# Subset the dataframe to remove genes with length <= threshold
genes_subset <- subset(gn_cds, len > threshold)

#rebin
genes_subset$decile <- ntile(genes_subset$oe_lof_upper, 10)#15738

#plot their 5utr length across deciles now
five_len <- subset(length_exon, select=c(gene_id,UTR_length))

#merge
gn_ten <- merge(five_len, genes_subset, by="gene_id")#16456

#plot 5utr length
ggplot(gn_ten, aes(x=factor(decile), y=UTR_length)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("UTR Length (bp)") 

#stats for 5'utr length still sig even when remvoing shortest genes
bot <- gn_ten %>% filter(decile<30)
mean(bot$UTR_length)#272.7194

top <-gn_ten %>% filter(decile>80)
mean(top$UTR_length)#171.8732
bot2 <- bot$UTR_length
top2 <- top$UTR_length
wilcox.test(bot2, top2)$p.value
#6.471138e-115


#replot cds length across deciles with shortest genes remvoed
#normal boxplot
ggplot(gn_ten, aes(x=factor(decile), y=len)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()


#plot 5'UTRs with shortest 10% genes gone as raincloud sitch

#new utr mean
mean(gn_ten$UTR_length)
#208.7699

gn_ten$decile <- gn_ten$decile *10

ggplot(data = gn_ten, aes(x=factor(decile), y = UTR_length, fill = decile)) +   geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8, fill="#793A92") +     geom_point(aes(y = UTR_length, fill = decile), position = position_jitter(width = 0.15), size = 1, alpha = 0.1, color="grey48") +     geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.5, fill="#793A92") +     labs(x = "LEOUF Decile", y = "UTR Length (bp)") + coord_cartesian(ylim=c(0, 2000)) + guides(fill = FALSE, color = FALSE) +theme_light()  + theme(text = element_text(size = 20)) + geom_hline(yintercept = 209,linetype = "dotted", size=0.8) +  scale_x_discrete(expand = expand_scale(add = c(0.2, 0.8)))

#11 genes cut off
```

#15. shuffled codons
```{r}
#shuffled codons in 5utrs to get o/e
#oe per gene id
#plot oe per gene
#merge gnoamd
#so mean codons oe per decile
#see imag ein shuf file from fred and re generate it

names(gn_final1)[10]<-"ensgid"
#get gene lists in bottom and top deciles
bot <- gn_final1 %>% filter(decile<3)#3832
top <- gn_final1%>% filter(decile>8)#3830

bot_gene <- subset(bot, select=c(ensgid))
top_gene <- subset(top, select=c(ensgid))

#merge with codon oe
oe_gene<- read.table("oe_gene.txt")
bot_oe <- merge(oe_gene, bot_gene, by="ensgid")
top_oe <- merge(oe_gene, top_gene, by="ensgid")



#do one manually to check maths
tat <- bot_oe %>% filter(codons=="TAT")
sum(tat$observed) #3275/3695
sum(tat$expected) #2625.92/3695
#not coming out the same.. 
#divide by gene total in group? to get mean
u <- unique(bot_oe$ensgid)#3695
#get same result which is diff when i get mean

#maybe sum all oberseved and expectd codons in that group then do mean?
botoe <- bot_oe %>% group_by(codons) %>% summarise(sum(observed), sum(expected))
#now i have how many of each codons was observed or expectd in this group
#get the oe for this group now
botoe$oe <- botoe$`sum(observed)`/botoe$`sum(expected)`
#diff to the other but i have more confidence in this one..

#do top
topoe <- top_oe %>% group_by(codons) %>% summarise(sum(observed), sum(expected))
#now i have how many of each codons was observed or expectd in this group
#get the oe for this group now
topoe$oe <- topoe$`sum(observed)`/topoe$`sum(expected)`

#plot 
botoe$group <- "LEOUF 1-3"
topoe$group <- "LEOUF 8-10"

plot_oe <- rbind(botoe, topoe)
ggplot(plot_oe, aes(x = reorder(codons, oe), y = oe, color = group)) + geom_point() + geom_hline(yintercept=1, linetype = "dotted")+facet_grid(group ~ .)+ theme(axis.text.x = element_text(angle = 90))
#what method is this using? if legit use this




#now get mean codon oe for each group (dont thin kthis is right but try and plot anywya tosee if there a is a diff?)
#botoe2 <- bot_oe %>% group_by(codons) %>% summarise(mean_oe=mean(oe))
topoe2 <- top_oe %>% group_by(codons) %>% summarise(mean_oe=mean(oe))
botoe2$group <- "LEOUF 1-3"
topoe2$group <- "LEOUF 8-10"
plotoe <- rbind(botoe2, topoe2)
ggplot(plotoe, aes(x = codons, y = mean_oe, color = group)) + geom_point() + facet_grid(group ~ .)+ theme(axis.text.x = element_text(angle = 90))


#4.5.23
#newer method for shuffled codons
o_e <- read.table("oe_mean_shuf.txt")
names(gn_final1)[10]<-"ensgid"
#get gene lists in bottom and top deciles
bot <- gn_final1 %>% filter(decile<3)#3832
top <- gn_final1%>% filter(decile>8)#3830
bot_gene <- subset(bot, select=c(ensgid))
top_gene <- subset(top, select=c(ensgid))

#change oe=nan
o_e[oe=="NaN", oe:=0]

bot_oe <- merge(o_e, bot_gene, by="ensgid")
top_oe <- merge(o_e, top_gene, by="ensgid")


#i want a mean oe per codon, with CIs
bot2 <- bot_oe %>% group_by(variable) %>% summarise(mean_oe=mean(oe))
top2 <- top_oe %>% group_by(variable) %>% summarise(mean_oe=mean(oe))

#bind and plot
bot2$group <- "LEOUF 1-3"
top2$group <- "LEOUF 8-10"
plot_shuf <- rbind(bot2, top2)



ggplot(plot_shuf, aes(x = reorder(variable, mean_oe), y = mean_oe, color = group)) + geom_point() + facet_grid(group ~ .)+geom_hline(yintercept=1, linetype = "dotted")+xlab("Codons")+ylab("o/e") + theme(axis.text.x = element_text(angle = 90))


#CIS
summary_oe_bot <- bot2[, .(
    median_oe= median(mean_oe, na.rm = TRUE), 
    lower_ci= quantile(mean_oe, prob=0.025, na.rm = TRUE),
    upper_ci= quantile(mean_oe, prob=0.975, na.rm = TRUE)
), by=.(variable)]

#not liking it, does it need to be a dt?


## try constrain to 5'utrs between 100-300bp and then do this. i want to see if ATG in higher leoufs have a lower value (appear less freq than higher leouf bec its a length thing)

shorter <- gnomad_len %>% filter(UTR_length <300)
#get bottom and top leouf deciles
tab <- as.data.frame(table(shorter$decile))
#way less in the lower deciles. wonder if once merged with mane we need to be re binning? originally was told no but not sure how thats not driving stuff too. redid with just less than 300 bp

bot <- shorter %>% filter(decile<3)
top <- shorter %>% filter(decile>8)

names(o_e)[1]<-"gene_id"
o_e <- as.data.table(o_e)
o_e[oe=="NaN", oe:=0]
o_e <- o_e %>% replace(is.na(.), 0)
bot_oe <- merge(o_e, bot, by="gene_id")
top_oe <- merge(o_e, top, by="gene_id")


#i want a mean oe per codon, with CIs
bot2 <- bot_oe %>% group_by(variable) %>% summarise(mean_oe=mean(oe))
top2 <- top_oe %>% group_by(variable) %>% summarise(mean_oe=mean(oe))

#bind and plot
bot2$group <- "LEOUF 1-3"
top2$group <- "LEOUF 8-10"
plot_shuf <- rbind(bot2, top2)



ggplot(plot_shuf, aes(x = reorder(variable, mean_oe), y = mean_oe, color = group)) + geom_point() + facet_grid(group ~ .)+geom_hline(yintercept=1, linetype = "dotted")+xlab("Codons")+ylab("o/e") + theme(axis.text.x = element_text(angle = 90))


```
#16. 5'UTR length as a proportion of whole mRNA length 
```{r}
#in response to a reviewer comment, get the 5'UTR length as a proportion of whole mRNA length and plot across LEOUF

#i already have CDS length, and 5'UTR length. get 3'UTR length

cds_len <- read.table("cds_len.txt")

### 3'UTR 
select_ensembl_genomic <- read.delim("MANE.GRCh38.v1.0.ensembl_genomic.gff", header=FALSE)

#i just need gene_id and lengths

three <- select_ensembl_genomic %>% filter(V3=="three_prime_UTR")
#split up last col
three_split <- cSplit(three, "V9", ";")
#only mane select transcripts
three2 <- three_split[grepl("MANE_Selec", three_split$V9_11), ]

#now calculate exon length per exon
three2$exon_len <- (three2$V5-three2$V4)+1
#aggregate to get total exon length per 3'UTR
#group by V9_03 and sum exon len column
three_final <- three2 %>% group_by(V9_03) %>% summarize(three_UTR_len = sum(exon_len))

#sort gene_id col
three_final <- str_remove_all(three_final$V9_03, "[gene_id=]") 
names(three_final)[1]<-"gene_id"

#get all in one df
length_exon <- read.table("length_introns_12.12.22.txt")

five_len <- subset(length_exon, select=c(gene_id,UTR_length))
#the three df have diff amounts of genes, how to merge them but add 0 ?
# Left join 5'UTR and coding sequence data frames
five_cds <- full_join(five_len, cds_len, by = "gene_id")

# Left join the merged data frame with the 3'UTR data frame
mrna <- full_join(five_cds, three_final, by = "gene_id")

# Replace NA values with 0
mrna[is.na(mrna)] <- 0

#now get full mrna len
mrna2 <- mrna %>% mutate(mrna_len = UTR_length + len + three_UTR_len)

#5'UTR as a percentage of full mrNA
mrna2$five_prop <- mrna$UTR_length/mrna$mrna_len
mrna2$five_perc <- (mrna$UTR_length/mrna$mrna_len) *100

#write.table(mrna2, "mrna_len_m1.txt")
mrna2 <- read.table("mrna_len_m1.txt")
#merge with gnomad and plot across louef
#split gene_id up
mrna2<- cSplit(mrna2, "gene_id", ".") 
mrna2 <- subset(mrna2, select=-c(gene_id_2))
names(mrna2)[7]<-"gene_id"


#####merge by gene id 
names(gn_final1)[10]<-"gene_id"
gnomad_len <- merge(mrna2, gn_final1, by = "gene_id") 

#plot
gnomad_len$decile <- gnomad_len$decile*10


ggplot(gnomad_len, aes(x=factor(decile), y=five_perc)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + theme(text = element_text(size = 23)) + xlab("LOEUF Decile") + ylab("Percentage") 

#want to plot side by side, full mRNA length, and 5'UTR as proportion one. na just plot seperately
ggplot(gnomad_len, aes(x=factor(decile), y=mrna_len)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + theme(text = element_text(size = 23))  + xlab("LOEUF Decile") 

#+ scale_y_continuous(limits = c(0, 1500))


#do stats
mrna2 <- read.table("mrna_len_m1.txt")

#run from above
bot <- gnomad_len %>% filter(decile<3)
mean(bot$five_perc)#0.0003802447

top <- gnomad_len %>% filter(decile>8)
mean(top$five_perc)#0.0002208087
bot2 <- bot$five_perc
top2 <- top$five_perc
wilcox.test(bot2, top2)$p.value
#1.245442e-167

#not sure can do wilcoxon on this..
t.test(bot2,top2)$p.value
# 1.255952e-85


#also did it on proportion not percentage (mimicing aug/length) - exactly same results

#run from above
bot <- gnomad_len %>% filter(decile<3)
mean(bot$five_prop)#0.0003802447

top <- gnomad_len %>% filter(decile>8)
mean(top$five_prop)#0.0002208087
bot2 <- bot$five_prop
top2 <- top$five_prop
wilcox.test(bot2, top2)$p.value
#1.245442e-167

#not sure can do wilcoxon on this..
t.test(bot2,top2)$p.value
#1.255952e-85
```

#17. cell essential/recessive genes
```{r}
# bring in the cell essential gene list from HArt et al but mcarthur github
#setwd("")
#bial <- read.table("macarthur biallelic list 1.7.21.txt")#1073
essential <- read.table("cell essential genes macarthur 8.11.21.txt")#683

#merge them with gnomad and plot 
#mainly need to plot the essential genes 
names(essential)[1]<-"gene"

ess_gn <- merge(gn_final1, essential, by="gene")
ess_gn$decile <- ess_gn$decile*10

tab_ess <- data.frame(table(ess_gn$decile))

ggplot(tab_ess, aes(x=factor(Var1), y=Freq)) + geom_bar(position="dodge", stat="identity") + theme_light()  + theme(text = element_text(size = 20)) + xlab("LOEUF Decile") + ylab("Gene Count") 

#stats - more cell essential genes in lower leouf deciles
tab_ess$Var1 <- as.numeric(as.character(tab_ess$Var1))
bot_ess <- tab_ess %>% filter(Var1<30)
top_ess <- tab_ess %>% filter(Var1>80)

#total of genes in bot and top 2 deciles
tot_bot <- sum(bot_ess$Freq)#254
tot_top <- sum(top_ess$Freq)#37

#total genes in deciles
total_gn <- data.frame(table(gn_final1$decile))
#total bottom
1916+1916
#3832

#total top
1915+1915
#3830


#total without essential genes
#bot
3832 - tot_bot
#3578

3830 -  tot_top
#3793

zero <- c(3578,3830)
ess <- c(tot_bot, tot_top)
ess_matrix <- data.frame(zero=zero,ess=ess)
chisq.test(ess_matrix)$p.value

#1.437978e-38

#percentages
(tot_bot/3832)*100
#6.628392
(tot_top/3830)*100
# 0.9660574

```